register.go

```diff
@@ -17,55 +17,148 @@ func getConsumerUUID() (string, error) {
+	if err != nil {
+		return orgs, err
+	}
+
+	for _, org := range organizations {
+		orgs = append(orgs, org.Key)
+	}
+
+	return orgs, nil
+}
+
+// registerUsernamePassword tries to register system against candlepin server (Red Hat Management Service)
+// username and password are mandatory. When organization is not obtained, then this method
+// returns list of available organization and user can select one organization from the list.
+func registerUsernamePassword(username, password, organization, serverURL string) ([]string, error) {
+	var orgs []string
 	if serverURL != "" {
 		if err := configureRHSM(serverURL); err != nil {
-			return fmt.Errorf("cannot configure RHSM: %w", err)
+			return orgs, fmt.Errorf("cannot configure RHSM: %w", err)
 		}
 	}
 
 	conn, err := dbus.SystemBus()
 	if err != nil {
-		return err
+		return orgs, err
 	}
 
 	uuid, err := getConsumerUUID()
 	if err != nil {
-		return err
+		return orgs, err
 	}
 	if uuid != "" {
-		return fmt.Errorf("warning: the system is already registered")
+		return orgs, fmt.Errorf("warning: the system is already registered")
 	}
 
 	registerServer := conn.Object("com.redhat.RHSM1", "/com/redhat/RHSM1/RegisterServer")
 
 	var privateDbusSocketURI string
-	if err := registerServer.Call("com.redhat.RHSM1.RegisterServer.Start", dbus.Flags(0), "").Store(&privateDbusSocketURI); err != nil {
-		return err
+	if err := registerServer.Call(
+		"com.redhat.RHSM1.RegisterServer.Start",
+		dbus.Flags(0),
+		"").Store(&privateDbusSocketURI); err != nil {
+		return orgs, err
 	}
 	defer registerServer.Call("com.redhat.RHSM1.RegisterServer.Stop", dbus.FlagNoReplyExpected, "")
 
 	privConn, err := dbus.Dial(privateDbusSocketURI)
 	if err != nil {
-		return err
+		return orgs, err
 	}
 	defer privConn.Close()
 
 	if err := privConn.Auth(nil); err != nil {
-		return err
+		return orgs, err
 	}
 
-	if err := privConn.Object("com.redhat.RHSM1", "/com/redhat/RHSM1/Register").Call("com.redhat.RHSM1.Register.Register", dbus.Flags(0), organization, username, password, map[string]string{}, map[string]string{}, "").Err; err != nil {
-		return unpackError(err)
+	if err := privConn.Object(
+		"com.redhat.RHSM1",
+		"/com/redhat/RHSM1/Register").Call(
+			"com.redhat.RHSM1.Register.Register",
+			dbus.Flags(0),
+			organization,
+			username,
+			password,
+			map[string]string{},
+			map[string]string{},
+			"").Err; err != nil {
+		rhsmError, _ := unpackError(err)
+		// When organization was not specified, and it is required to specify it, then
+		// try to get list of available organizations
+		if organization == ""  && rhsmError.Exception == "OrgNotSpecifiedException" {
+			orgsCall := privConn.Object(
+				"com.redhat.RHSM1",
+				"/com/redhat/RHSM1/Register",
+			).Call(
+				"com.redhat.RHSM1.Register.GetOrgs",
+				dbus.Flags(0),
+				username,
+				password,
+				map[string]string{},
+				"",
+			)
+			var s string
+			err = orgsCall.Store(&s)
+			if err == nil {
```

Comment:
Can you flip the logic around this to `if err != nil`? The rest of the program uses that pattern for error checking, and I think maintaining a consistent pattern is important.

Reply:
Sure, I can.